src/main/java/org/jitsi/jibri/api/xmpp/XmppApi.kt

```diff
@@ -0,0 +1,214 @@
+                    JibriStatusPacketExt.Status.IDLE,
+                    JidCreate.bareFrom("${config.controlMuc.roomName}@${config.controlMuc.domain}"))
+                mucClient.sendStanza(jibriStatus)
+            }
+        }
+    }
+
+    /**
+     * Helper function to handle a [JibriIq] message with the context of the [XmppEnvironmentConfig] and [MucClient]
+     * that this [JibrIq] was received on.
+     */
+    private fun handleJibriIq(jibriIq: JibriIq, xmppEnvironment: XmppEnvironmentConfig, mucClient: MucClient): IQ {
+        logger.info("Received JibriIq $jibriIq from environment ${xmppEnvironment.name}")
+        return when (jibriIq.action) {
+            JibriIq.Action.START -> handleStartJibriIq(jibriIq, xmppEnvironment, mucClient)
+            JibriIq.Action.STOP -> handleStopJibriIq(jibriIq)
+            else -> IQ.createErrorResponse(
+                jibriIq,
+                XMPPError.getBuilder().setCondition(XMPPError.Condition.bad_request))
+        }
+    }
+
+    /**
+     * Handle a start [JibriIq] message.  We'll respond immediately with a [JibriIq.Status.PENDING] IQ response and
+     * send a new IQ with the subsequent stats after starting the service:
+     * [JibriIq.Status.FAILED] if there was an error starting the service (or an error while the service was running)
+     * [JibriIq.Status.BUSY] if the Jibri was already busy and couldn't start the requested service
+     * [JibriIq.Status.ON] if the service started successfully
+     */
+    private fun handleStartJibriIq(
+            startJibriIq: JibriIq,
+            xmppEnvironment: XmppEnvironmentConfig,
+            mucClient: MucClient): IQ {
+        logger.info("Received start request")
+        // We don't want to block the response to wait for the service to actually start, so submit a job to
+        // start the service asynchronously and send an IQ with the status after its done.
+        executor.submit {
+            val resultIq = JibriIq()
+            resultIq.to = startJibriIq.from
+            resultIq.type = IQ.Type.set
+            try {
+                logger.info("Starting service")
+
+                // If there is an issue with the service while it's running, we need to send an IQ
+                // to notify the caller who invoked the service of its status, so we'll listen
+                // for the service's status while it's running and this method will be invoked
+                // if it changes
+                val serviceStatusHandler: JibriServiceStatusHandler = { serviceStatus ->
+                    when (serviceStatus) {
+                        JibriServiceStatus.ERROR -> {
+                            logger.info("Current service had an error")
+                            val errorIq = JibriIq()
+                            errorIq.to = startJibriIq.from
+                            errorIq.type = IQ.Type.set
+                            errorIq.status = JibriIq.Status.FAILED
```

Comment:
I'm seeing this pattern a lot where JibriIq() creates something and then you manually set values on it. Is this stylistically how attributes are normally set? I'd be more used to passing these into the factory function to potentially hide some kind of implementation, if it even needs hiding.

Reply:
unfortunately there aren't many constructors in JibriIq.  I already added a helper function for creating an iq response, i think it would be nice to add a helper for this case too.